### PR TITLE
Metrics collect stress test

### PR DESCRIPTION
**Pull request by fraillt**
## Changes

Added two tests (delta and cumulative temporality) to measuring how much metrics collection phase impact measurements throughput.
Implementation consists of running collection phase in the loop, while simultaneously testing measurements throughput.
The main concern isn't the speed of collection phase itself, but ensuring that Opentelemetry Metrics doesn't significantly contribute to p99 latency.
I didn't customize it for different metrics type, idea is to eventually land #2117, so that I could unify collection phase for other metrics #2145.

Additionally, significantly improved code reuse for `stress` crate.

Here are some results on my machine:
```
cargo run --release --package stress --bin metrics_collect -- delta
    Finished `release` profile [optimized] target(s) in 0.07s
     Running `target/release/metrics_collect delta`
     17.38 measurements/ms
     52.54 measurements/it
   3023.88 μs/it
   
cargo run --release --package stress --bin metrics_collect -- cumulative
    Finished `release` profile [optimized] target(s) in 0.07s
     Running `target/release/metrics_collect cumulative`
      9.14 measurements/ms
     24.35 measurements/it
   2665.30 μs/it
```   

For curiosity I have also implemented #2117 and #2145 locally, and results are as follows:
```
cargo run --release --package stress --bin metrics_collect -- delta
    Finished `release` profile [optimized] target(s) in 0.07s
     Running `target/release/metrics_collect delta`
     56.38 measurements/ms
    154.24 measurements/it
   2735.73 μs/it
   
cargo run --release --package stress --bin metrics_collect -- cumulative
    Finished `release` profile [optimized] target(s) in 0.07s
     Running `target/release/metrics_collect cumulative`
    840.84 measurements/ms
   2077.64 measurements/it
   2470.90 μs/it   
```

## Merge requirement checklist

* [x] [CONTRIBUTING](https://github.com/open-telemetry/opentelemetry-rust/blob/main/CONTRIBUTING.md) guidelines followed
* [ ] Unit tests added/updated (if applicable)
* [ ] Appropriate `CHANGELOG.md` files updated for non-trivial, user-facing changes
* [ ] Changes in public API reviewed (if applicable)

### CODE REVIEW

**Comment by utpilla** [comment]
@fraillt Thanks for the continued interest in making metrics better! I have two suggestions around sending PRs that I would like you to consider. These are mainly directed towards making the PR review process easy and simple (which would in turn result in the getting PRs merged faster).

1. Send targeted PRs aimed at solving one thing at a time. For example, there's a lot of things going on this PR. There's a new test being added as the PR title suggests, there is moving around of files, refactoring around reuse of logic across projects. Each of these different additions/enhancements could be added incrementally instead of doing it in a single shot.

2. Avoid force pushing the commit. With force push, we have no way to know how many files changed since your last commit. We then have to review all the files again to ensure that the latest commit only made the changes that we were expecting. Retaining the individual commits makes it easy for reviewer.

**Comment by fraillt** [comment]
> @fraillt Thanks for the continued interest in making metrics better! I have two suggestions around sending PRs that I would like you to consider. These are mainly directed towards making the PR review process easy and simple (which would in turn result in the getting PRs merged faster).
> 
>     1. Send targeted PRs aimed at solving one thing at a time. For example, there's a lot of things going on this PR. There's a new test being added as the PR title suggests, there is moving around of files, refactoring around reuse of logic across projects. Each of these different additions/enhancements could be added incrementally instead of doing it in a single shot.
> 
>     2. Avoid force pushing the commit. With force push, we have no way to know how many files changed since your last commit. We then have to review all the files again to ensure that the latest commit only made the changes that we were expecting. Retaining the individual commits makes it easy for reviewer.

I'll try to be more careful regarding force-push in the future.
Regarding targeted PR, I see your point, that targeted PRs are easier to review, but sometimes they doesn't make sense in isolation. For instance I moved `SharedReader` from `opentelemetry-sdk/benches/metric.rs` to `opentelemetry-sdk/src/testing/metrics/metric_reader.rs`, in isolation that looks like a pointless effort, but I do need this functionality for `metrics_collect.rs`.

Basically what I'm trying to say, is that I'm afraid that if I create PR just for refactoring, it might look worthless on its own...

All changes that was done in this PR does relate to one file that brings new functionality `stress/src/bin/metrics_collect.rs`.

**Comment by utpilla** [comment]
> Basically what I'm trying to say, is that I'm afraid that if I create PR just for refactoring, it might look worthless on its own...
> 
> All changes that was done in this PR does relate to one file that brings new functionality `stress/src/bin/metrics_collect.rs`.

I think we could still have a simpler split up of PRs. Something like this:
1. Send a PR that adds the new stress test. (along with the boiler plate code: `SharedReader` etc.) You don't have to add the reuse related changes at this point.
2. Once that gets merged, you could send a refactoring PR to allow for reuse of `SharedReader`. This would showcase the motivation for reuse as well.

**Comment by fraillt** [comment]
I force-pushed, because it basically delete everything I did, except for one file, so I guess it is ok :).
I'll not force-push, once someone reviews and I'll update my revision to address review comments.

File: stress/src/metrics_collect.rs [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+                handle.join().unwrap();
+            }
+
+            // simultaneously start collecting and creating more measurements
+            for _ in 0..num_threads - 1 {
+                s.spawn(|| {
+                    barrier.wait();
+                    let now = Instant::now();
+                    let mut count = 0;
+                    while is_collecting.load(Ordering::Acquire) {
```
Comment (utpilla) [inline stress/src/metrics_collect.rs:135]: I think we could have a much simpler and effective setup here. If we want to know whether running collect stops the world, it's better to have spawn a thread that keeps calling `reader.collect` on a `loop`. And have other threads, record measurements simultaneously.

File: stress/src/metrics_collect.rs [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+    let start = Instant::now();
+    while start.elapsed() < Duration::from_secs(3) {
+        res.num_iterations += 1;
+        let is_collecting = AtomicBool::new(false);
+        let measurements_while_collecting = AtomicUsize::new(0);
+        let time_while_collecting = AtomicUsize::new(0);
+        let barrier = Barrier::new(num_threads + 1);
+        std::thread::scope(|s| {
+            // first create bunch of measurements,
+            // so that collection phase wouldn't be "empty"
```
Comment (utpilla) [inline stress/src/metrics_collect.rs:108]: This can be done without spawning new threads. (before you spawn threads for running collect and recording measurements)

File: stress/src/metrics_collect.rs (hunk 1 above)
Comment (fraillt) [reply stress/src/metrics_collect.rs:135]: I did that initially. The problem was that collection phase was not realistic as it had 0 measurements and basically held lock in a loop. So I decided to make it more realistic, by generating some measurements so collection wouldn't be empty.

Comment (utpilla) [reply stress/src/metrics_collect.rs:135]: > The problem was that collection phase was not realistic as it had 0 measurements and basically held lock in a loop

I don't follow. You could always record some measurements before you call collect. My concern is around the way this test is setup. The test is using some custom "iterations" which seems unnecessary. You could simply do this instead:
1. Emit some measurements.
2. Spawn a thread that runs collect in a loop.
3. Spawn additional threads that record measurements.
4. Calculate throughput for the measurements recorded in step 3.

If doing the above, leads to zero measurements being recorded, then we have our answer: `Collect is going to "stop the world" as long as it runs`.

As I mentioned in this [comment](https://github.com/open-telemetry/opentelemetry-rust/pull/2247#discussion_r1823416061), unless we plan to use a more efficient synchronization mechanism in `ValueMap`, this stress test would not be adding any value.

Comment (fraillt) [reply stress/src/metrics_collect.rs:135]: First of all, this test is adding value, as it is able to measure difference with different "collect" implementation.
Second, running collect in the loop doesn't work with delta temporality, because after first run there will be zero measurements and all other iterations will basically hold write lock in the loop. While in reality no one runs collect in loop and usually there are many measurements and its important that write lock is held as short as possible (which is not the case with current implementation, hence I was able to improve it and this test actually measures it).
I could probably have different tests for delta an cumulative temporality, but I decided to emulate realistic environment.

File: stress/src/metrics_collect.rs [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+            }
+
+            let collect_handle = s.spawn(|| {
+                let mut rm = ResourceMetrics {
+                    resource: Resource::empty(),
+                    scope_metrics: Vec::new(),
+                };
+                is_collecting.store(true, Ordering::Release);
+                barrier.wait();
+                reader.collect(&mut rm).unwrap();
```
[the remark this answers is not in the document]
Comment (fraillt) [reply stress/src/metrics_collect.rs:160]: For delta temporality I was able to reduce the amount of time a write lock is held while collecting.
For cumulative temporality I simply changed to read lock (currently there is write lock, although cumulative temporality doesn't modify anything, and IMO locking individual measurements is sufficient and much better, then "stop the world until all measurements are read" approach). Although read lock is better, but measurements with new attribute-sets will still be locked. I see two more alternatives: 
* clone entire attribute-set hashmap (acquire read lock, clone hashmap, release lock, iterate individual measurements)
* implement some sort of sharding (more throughtput at the cost of single measurement performance)

In any case, there are ways to improve it:) so I created this PR so we could actually measure different ideas.

File: stress/src/metrics_collect.rs (hunk 1 above)
Comment (utpilla) [reply stress/src/metrics_collect.rs:135]: > While in reality no one runs collect in loop

In reality, no one runs the collect the way it's run in this stress test either. Collect would usually be run periodically in 10/30/60 seconds intervals.

> First of all, this test is adding value, as it is able to measure difference with different "collect" implementation.

> usually there are many measurements and its important that write lock is held as short as possible (which is not the case with current implementation, hence I was able to improve it and this test actually measures it).

This test [claims](https://github.com/open-telemetry/opentelemetry-rust/pull/2247/files#diff-7cd495a1f764caabee463e9e132e20d67bcf0de3da1ea032c6baf39fcf8dbe9bR44-R45) to measure update throughput "while collect is running". The actual implementation of this test however relies on squeezing in some updates around collect (either before collect executes or after it). That's not the same as testing "while collect is running".

I would love to improve collect to take the write lock for the shortest possible time, but it's better tested using a benchmark.

File: stress/src/metrics_collect.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (utpilla) [reply stress/src/metrics_collect.rs:160]: These are good ideas!

> For cumulative temporality I simply changed to read lock (currently there is write lock, although cumulative temporality doesn't modify anything, 

Great! In that case, we should be able test update throughput while `collect` runs in a loop for `Cumulative`. 🙂

[the remark this answers is not in the document]
Comment (lalitb) [reply stress/src/metrics_collect.rs:160]: > implement some sort of sharding (more throughtput at the cost of single measurement performance)

I did try sharding approach earlier in #1564 - to reduce the write-local duration by locking only the specific shard of measurements during collection. However, I encountered some concurrency issues. It’s still an approach worth revisiting at some point.

File: stress/src/metrics_collect.rs (hunk 1 above)
Comment (fraillt) [reply stress/src/metrics_collect.rs:135]: >  The actual implementation of this test however relies on squeezing in some updates before collect runs.
There's a lot of truth to this statement, but none the less it appears to measure something useful  :)

| temporality | type of change                               | before change (measurements/ms) | after change (measurements/ms) |
|-------------|----------------------------------------------|---------------------------------|--------------------------------|
| Cumulative  | changed write to read lock of hashmap        | 9                               | 840                            |
| Delta       | reduce the amount of time write lock is held | 17                              | 56                             |

Ideally I would like to "catch" changes to how attribute-set hashmap are locked with both: existing types of measurements (with existing attribute-set and new attribute-set combination).
Maybe we need to tests these things separately... I don't know, any ideas are welcome :)

**Comment by fraillt** [comment]
#2323 solves same problem in much more precise way.
